plugins/outputs/exec/exec.go

```diff
@@ -62,23 +64,39 @@ func (e *Exec) Close() error {
 
 // Write writes the metrics to the configured command.
 func (e *Exec) Write(metrics []telegraf.Metric) error {
+	e.outBuffer.Reset()
 	var buffer bytes.Buffer
-	serializedMetrics, err := e.serializer.SerializeBatch(metrics)
-	if err != nil {
-		return err
-	}
-	buffer.Write(serializedMetrics)
+	if e.UseBatchFormat {
+		serializedMetrics, err := e.serializer.SerializeBatch(metrics)
+		if err != nil {
+			return err
+		}
+		buffer.Write(serializedMetrics)
 
-	if buffer.Len() <= 0 {
-		return nil
+		if buffer.Len() <= 0 {
+			return nil
+		}
+
+		return e.runner.Run(time.Duration(e.Timeout), e.Command, e.Environment, &buffer, &e.outBuffer)
 	}
+	errs := make([]error, 0, len(metrics))
+	for _, metric := range metrics {
+		serializedMetric, err := e.serializer.Serialize(metric)
+		if err != nil {
+			return err
+		}
+		buffer.Reset()
+		buffer.Write(serializedMetric)
 
-	return e.runner.Run(time.Duration(e.Timeout), e.Command, e.Environment, &buffer)
+		err = e.runner.Run(time.Duration(e.Timeout), e.Command, e.Environment, &buffer, &e.outBuffer)
+		errs = append(errs, err)
```

Comment:
Maybe return at first error?

Reply:
Such that all the rest of the metrics in the batch aren't sent after the first error?

Reply:
Indeed.

Reply:
WDYT?

Reply:
My inclination is to keep running on the other metrics, but I don't have a strong preference.

Reply:
I'm not sure telegraf does do a retry, but if it does, it would resend all of them to the output..

Reply:
Here the question is a bit on if the error is retryable... If for example the script encounters a temporary issue (disk-full, network down etc) it would make sense to **return an error** forcing Telegraf to retry **the whole batch**. Otherwise, e.g. on wrong credentials, format errors etc, we don't want a retry.
For now I would return the error and let Telegraf do the retry as that's what the plugin also does up to now. However, it might make sense to add another option (in another PR) to control that behavior as we cannot know what to do.